### ISSUE TITLE
SAGA: Cannot read property cancel of undefined

### ISSUE THREAD

**Issue by aloysius-tim** [react-boilerplate/react-boilerplate#2314]
# React Boilerplate

## Issue Type

- [X] Bug (https://github.com/react-boilerplate/react-boilerplate/blob/master/.github/CONTRIBUTING.md#bug-reports)
- [ ] Feature (https://github.com/react-boilerplate/react-boilerplate/blob/master/.github/CONTRIBUTING.md#feature-requests)

## Description

Hi there,

Just wanted to point out that the following issue has not been resolved yet: https://github.com/react-boilerplate/react-boilerplate/issues/2021

@kai23 solution of adding '&& descriptor.task' on line 75 worked for me.

```
export function ejectSagaFactory(store, isValid) {
  return function ejectSaga(key) {
    if (!isValid) checkStore(store);

    checkKey(key);

    if (Reflect.has(store.injectedSagas, key)) {
      const descriptor = store.injectedSagas[key];
      if (descriptor.mode && descriptor.mode !== DAEMON && descriptor.task) {
        descriptor.task.cancel();
        // Clean up in production; in development we need `descriptor.saga` for hot reloading
        if (process.env.NODE_ENV === 'production') {
          // Need some value to be able to detect `ONCE_TILL_UNMOUNT` sagas in `injectSaga`
          store.injectedSagas[key] = 'done'; // eslint-disable-line no-param-reassign
        }
      }
    }
  };
}
```

**Comment by julienben** [comment]
Hey @aloysius-tim,

Thanks for submitting this. I've never seen it on my end nor do I know how to reproduce it.

Do you have any idea why adding this check fixes it though?

Read: Does this simply bypass the bug or is it a legitimate check?

Cheers!

**Comment by NathanBWaters** [comment]
I just hit this error.  Just want to note that I'm only hitting this in production mode and not development mode.  Adding the `&& descriptor.task` fix also worked for me.

**Comment by tatemz** [comment]
> Note: I don't have a suggested fix, but hopefully the below details can help the next dev better understand the issue.

I too have this issue in `production` mode. The issue occurs when I switch languages using `react-intl`. Some of my components use `InjectIntl` directly in order to translate strings (rather than HTML elements). See [API docs](https://github.com/yahoo/react-intl/wiki/API#injection-api) for more details.

The root of the problem lies within the React component lifecycle.

`InjectIntl` will wrap my component. `injectSaga` also wraps my component. Since the `ejectSaga` logic is run on `componentWillUnmount`, the logic is asynchronous. In my particular case, changing the language within my app triggers an update down to my component that is wrapped in both `InjectIntl` and `injectSaga`.

The resulting order of execution is as follows:

1. App loads
2. Wrapped component mounts
3. Saga is injected with the should-be unique `key`

Debugging output at this point is as follows:
```
componentWillMount homePage InjectIntl(n)
{
    homePage: {saga: ƒ, mode: "@@saga-injector/restart-on-remount", task: {…}}
}
```

4. User interaction causes dispatched action to update the selected language/locale
5. The updated `InjectIntl` wrapped component mount logic applies *FIRST* (which re-injects the saga)

Debugging output for this action is as follows:
```
componentWillMount homePage InjectIntl(n)
{
    homePage: {saga: ƒ, mode: "@@saga-injector/restart-on-remount", task: {…}}
}
```

At this point, you may be able to see the obvious problem... the saga was already injected (and had not yet ejected), so re-injecting the already existing saga does nothing.

6. The _older_ `InjectIntl` wrapped component unmount logic applies *SECOND* (which ejects the saga)

Debugging output for this action is as follows:
```
componentWillUnmount homePage InjectIntl(n)
{
    homePage: "done"
}
```

From this point , any attempt to run the saga will result in the following exception:
```
uncaught at Ne TypeError: Cannot read property 'cancel' of undefined
```

**Comment by flameddd** [comment]
@tatemz  I just face similar scenario with your few day ago. I elaborate here maybe could be someone ref.

In my case, when `container/App/index.js` mounted, `componentDidMount` will `dispatch(initApp())` action to fetch `userConfig`. And we `set language locale` According `userConfig`, and things happen

According to my understanding：
1. `saga A(1)` is injected for mount `Page A(1)`
2. `dispatch(initApp())`
3. `set language locale`
4. `<IntlProvider />` re-mount (due to `react-boilerplate` design `locale` as `element's key`)
5. make `Page A` remount

(and due to `React 16 life change`)

6. `saga A(2)` is injected for `Page A(2)` WillMount
7. `saga A(1)` is ejected for mount `Page A(1)` WillUnMount

at this point, `saga A(2)` still work perfectly.

and We find out things broken by following scenario

8. push to `Page B` (`saga A(2)` didn't cancel, cus `redux-saga` think it not running. without any console.error)
9. push to `Page A` then `saga A(3)` injected
10. then every dispatch(action()) will make trigger twice

## react-intl
further explain, In my view, I don't think this design is a bug
 > `react-boilerplate` design `locale` as `element's key`


## redux-saga
```javascript
// /node_modules/redux-saga/es/internal/proc.js

  function cancel() {
    // at step 8
   console.log(iterator._isRunning) // false
   console.log(!iterator._isCancelled) // false
    if (iterator._isRunning && !iterator._isCancelled) {
      iterator._isCancelled = true;
      taskQueue.cancelAll();

      end(TASK_CANCEL);
    }
  }
```

FYI

**Comment by julienben** [comment]
@tatemz Thanks for the detailed explanation. I'm very interested in fixing this but I'm having trouble following your explanation all the way through as I don't need / have removed the internationalization feature in my RBP-based projects. 

1. Any chance you could provide a set of instructions on how to reproduce the bug? 

2. Did you try the `&& descriptor.task` suggestion and did it work for you?

My purpose here is to fix the real/underlying issue while not creating any regressions.

Many thanks for future help!! ;)

**Comment by julienben** [comment]
As you can see in #2021, this has been around for a while. As have the suggested fixes. Let's get to the bottom of this...

**Comment by tatemz** [comment]
I can in a day or so, however the root of the issue in my case is that `react-intl` causes the route to re-render.

From [#2021](https://github.com/react-boilerplate/react-boilerplate/issues/2021#issuecomment-360681876)
> In react-router's Switch, mount and unmount are not insured to called in some specific order.

**Comment by julienben** [comment]
(Just tagging #2266 which might not be directly related but is also connected to issues in the saga injection setup.)

**Comment by julienben** [comment]
@flameddd @tatemz @aloysius-tim 

Would you mind please checking if the changes we made in PR #2468 have fixed this issue? Then please report back so we can finally mark this as resolved :)

**Comment by kwent** [comment]
Personally https://github.com/react-boilerplate/react-boilerplate/pull/2468 didn't help. but adding `&& descriptor.task` did.

**Comment by michaelcuneo** [comment]
This also happened constantly in my projects using RB, pre 3.7.0. I have not had this occur since 3.7.0. I am unable to determine the change that fixed it between the old version and new. If I see it, I'll contribute. But I have a feeling that it was directly related to Component vs PureComponent usage in my case.

**Comment by vospascal** [comment]
i did the following in injectSaga


    componentWillMount() {
      const { injectSaga, ejectSaga } = this.injectors;
      const injectedArgs = args || [this.props];
      const { store } = this.context;

      // eject old saga with same name, so it does not get double sagas injected
      if (
        Reflect.has(store.injectedSagas, key)
        && store.injectedSagas[key].saga === saga
      ) {
        ejectSaga(key);
      }

      injectSaga(key, { saga, mode }, ...injectedArgs);
    }

never have duplicated injected sagas this way

**Comment by imagine10255** [comment]
@vospascal 
I tried to use this method after
Route.push PrePage Saga disappears

**Comment by vospascal** [comment]
@imagine10255 try https://github.com/supasate/connected-react-router/blob/master/FAQ.md

**Comment by imagine10255** [comment]
@vospascal Sorry, I mean, I have a registration form for step1 ~ steo4 (have the function of returning to the previous step)

- Step1
  - Press to send redux action->onSubmit->saga call->api->route-push Step2
- Step2
  - Press back to the previous page (Route push Step1)
- Step1
  - Press to send redux action->onSubmit->.. {miss saga}

**Comment by praveen612** [comment]
I ran into similar issue, and this was because of race condition between injecting saga and ejecting them. 
componentWillUnMount is used for ejecting the saga which is a asyncrounous in React 16. 
Because of this you may inject a new instance of saga to the same key making the previous saga unaccessible. This should be fixed by keeping sagas mode DAEMON by default and should be fixed with #2430

**Comment by Apak00** [comment]
@metapraveen , yeah in my case it was 2 components trying to inject different sagas to same key when routing from one to another

> I ran into similar issue, and this was because of race condition between injecting saga and ejecting them.
> componentWillUnMount is used for ejecting the saga which is a asyncrounous in React 16.
> Because of this you may inject a new instance of saga to the same key making the previous saga unaccessible. This should be fixed by keeping sagas mode DAEMON by default and should be fixed with #2430

**Comment by julienben** [comment]
Closing for lack of activity + the APIs have changed so much I'm not sure these issues are still relevant. If you need to report new bugs with reducer or saga injection, please open an issue in our sister project [redux-injectors](https://github.com/react-boilerplate/redux-injectors). Thanks 🙏
